addons/l10n_in/models/account_invoice.py

```diff
@@ -77,6 +78,49 @@ def _onchange_name_warning(self):
             }}
         return super()._onchange_name_warning()
 
+    @api.depends('invoice_line_ids.product_id', 'company_id.l10n_in_hsn_code_digit')
+    def _compute_hsn_code_warning(self):
+
+        def build_warning(record, action_name, message, views, domain=False):
+            return {
+                'message': message,
+                'action_text': _("View %s", action_name),
+                'action': record._get_records_action(name=_("Check %s") % action_name, target='current', views=views, domain=domain or [])
+            }
+
+        indian_invoice = self.filtered(lambda m: m.country_code == 'IN')
+        for move in indian_invoice:
+            if move.company_id.l10n_in_hsn_code_digit and move.invoice_line_ids.filtered(lambda line: line.display_type == 'product' and line.tax_ids):
+                new_lines = lines = self.env['account.move.line']
+                for line in move.invoice_line_ids.filtered(lambda line: line.display_type == 'product' and line.tax_ids):
+                    if (line.l10n_in_hsn_code and (len(line.l10n_in_hsn_code) not in [4, 6, 8] or len(line.l10n_in_hsn_code) < int(move.company_id.l10n_in_hsn_code_digit))) or not line.l10n_in_hsn_code:
+                        lines |= line
+
+                for line in lines:
+                    if line._origin:
+                        new_lines |= line._origin
+
+                digit_suffixes = {
+                    '4': _("4 digits, 6 digits or 8 digits"),
+                    '6': _("6 digits or 8 digits"),
+                    '8': _("8 digits")
+                }
+                msg = _("Ensure that the HSN/SAC Code consists either {} in product line").format(
+                    digit_suffixes.get(move.company_id.l10n_in_hsn_code_digit, _("Invalid HSN/SAC Code digit"))
+                )
+                move.l10n_in_hsn_code_warning = {
+                    'invalid_hsn_code_length': build_warning(
+                        message=msg,
+                        action_name=_("Journal Items(s)"),
```

Comment:
@zepa-odoo @ajjo-odoo Can we make 2 warnings here: one for the lines with no product (or where product hsn != line hsn and line hsn not valid) which goes to journal items as here and one with products (when product hsn == line hsn and hsn not valid) which goes to a product view?

Reply:
@jco-odoo what i do believe and expect with this task is,
we can keep warning on invoice based on precision selection available in settings.
let's keep the warning over a product as it is.

Reply:
@jco-odoo 
according to me it's fine to have one warning. checking from product.
as we set hsn requirement over a product so i do beleive that , line item without product shall not considered for warning.
practically no one will operate accounting with hsn without product.

cc- @zepa-odoo

Reply:
Thanks for the feedback, @jco-odoo. 
I've updated the code based on @ajjo-odoo's comments. 

Please review the changes.